shard/engine.go

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2018 The CovenantSQL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package shard
+
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/CovenantSQL/CovenantSQL/utils/log"
+	"github.com/CovenantSQL/sqlparser"
+	"github.com/pkg/errors"
+)
+
+func getShardTS(sqlVal *sqlparser.SQLVal, args []driver.NamedValue) (insertTS int64, err error) {
+	var (
+		insertTime time.Time
+		argIndex   int64
+		arg        *driver.NamedValue
+	)
+
+	switch sqlVal.Type {
+	case sqlparser.StrVal:
+		insertTime, err = ParseTime(string(sqlVal.Val))
+		if err != nil {
+			return -1,
+				errors.Wrap(err, "unsupported: time column")
+		}
+		insertTS = insertTime.Unix()
+	case sqlparser.IntVal:
+		insertTS, err = strconv.ParseInt(string(sqlVal.Val), 10, 64)
+		if err != nil {
+			return -1,
+				errors.Wrap(err, "unsupported: time int val")
+		}
+
+	case sqlparser.ValArg:
+		if len(sqlVal.Val) > 1 {
+			for _, a := range args {
+				// Name has higher priority
+				if a.Name == string(sqlVal.Val[1:]) {
+					arg = &driver.NamedValue{
+						Name:    a.Name,
+						Ordinal: a.Ordinal,
+						Value:   a.Value,
+					}
+					break
+				}
+			}
+			if arg == nil {
+				argIndex, err = getValArgIndex(sqlVal)
+				if err != nil {
+					log.Warn(err)
+				} else {
+					for _, a := range args {
+						if a.Ordinal == int(argIndex) {
+							arg = &driver.NamedValue{
+								Name:    a.Name,
+								Ordinal: a.Ordinal,
+								Value:   a.Value,
+							}
+							break
+						}
+					}
+				}
+			}
+		}
+
+		if arg == nil {
+			return -1,
+				errors.Errorf("unsupported: %s named args",
+					string(sqlVal.Val))
+		}
+
+		switch arg.Value.(type) {
```

Comment:
Can be replaced with `switch v := arg.Value.(type)` and use `v` directly in following statements.